scarif/rail_control.py

```diff
@@ -0,0 +1,403 @@
+###############################################################################
+# Copyright (c) 2019, National Research Foundation (Square Kilometre Array)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+
+"""Convenience classes for sending Grbl over a serial port from python"""
+import re
+import logging
+import time
+
+from serial import Serial
+from collections import OrderedDict
+
+
+logging.basicConfig(level=logging.DEBUG,
+                    format='%(asctime)s %(levelname)s %(message)s',
+                    handlers=[logging.StreamHandler()])
+logger = logging.getLogger(__name__)
+
+
+class GrblSerialException(Exception):
+    """Grbl serial communications error"""
+    pass
+
+
+class GrblTimeoutException(Exception):
+    """A grbl timeout on sending a command"""
+    pass
+
+
+class GrblCommandException(Exception):
+    """A grbl command error"""
+    pass
+
+
+class GrblMotionModeExeption(Exception):
+    """A grbl motion command exception"""
+    pass
+
+
+class GrblCtrl(object):
+    """Low level abstraction of grbl commands as class methods
+
+    Parameters
+    ----------
+    port : string : USB serial device to connect to, e.g. /dev/serial
+    baudrate: integer : communications bit rate, e.g. 115200
+    timeout: integer : USB serial communcations timeout in seconds
+
+    Raises
+    ------
+    GrblSerialException: The serial port is not open
+    GrblCommandException: Basic command checking before sending
+    """
+    def __init__(self, port, baudrate, timeout):
+        super(GrblCtrl, self).__init__()
+        self._serial = Serial()
+        self._serial.port = port
+        self._serial.baudrate = baudrate
+        self._serial.timeout = timeout
+
+    def open(self):
+        self._serial.open()
+
+    def close(self):
+        self._serial.close()
+
+    def _command(self, command):
+        """Wrapper for executing a grbl command"""
+        if not self._serial.isOpen():
+            raise GrblSerialException('%s port is not open' % (self.port))
+        result = self._serial.write(command+'\n')
+        logger.debug('cmd: "%s" ret %i' % (command, result))
```

Comment:
FYI: It is generally better to do the following:
logger.debug('cmd: "%s" ret %i', command, result)

Pass in all the variables to the debug method and it will do the string processing, but only if it is needed.